## Exploring Hacker News Posts



Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

You can find the data set [HERE](https://www.kaggle.com/hacker-news/hacker-news-posts), but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions. Below are descriptions of the columns:

| Column | Description |
|:--|:--|
|0 - id | The unique identifier from Hacker News for the post |
|1 -  title | The title of the post |
|2 -  url | The URL that the posts links to, if it the post has a URL |
|3 -  num_points | The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes |
|4 -  num_comments | The number of comments that were made on the post |
|5 -  author | The username of the person who submitted the post |
|6 -  created_at | the date and time the post was made (the time zone is Eastern Time in the US) |

We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. Below are a couple examples:

 - Ask HN: How to improve my personal website?
 - Ask HN: Am I the only one outraged by Twitter shutting down share counts?
 - Ask HN: Aby recent changes to CSS that broke mobile?

Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. Below are a couple of examples:

 - Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'
 - Show HN: Something pointless I made
 - Show HN: Shanhu.io, a programming playground powered by e8vm

We'll compare these two types of posts to determine the following:

* Do Ask HN or Show HN receive more comments on average?
* Do posts created at a certain time receive more comments on average?

In [1]:
import csv

hn = list(csv.reader(open('hacker_news.csv')))

hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

We'll remove the headers first to isolate the data and make it easier to work with going forward. 

In [2]:
headers = hn[0]
hn = hn[1:]

hn[:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

In [3]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print('Ask HN length:', len(ask_posts))
print('Show HN length:', len(show_posts))
print('Other posts:', len(other_posts))
print('\n')
print('First 5 of Ask HN list')
print(ask_posts[:5])
print('\n')
print('First 5 of Show HN list')
print(show_posts[:5])


Ask HN length: 1744
Show HN length: 1162
Other posts: 17194


First 5 of Ask HN list
[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


First 5 of Show HN list
[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'],

We've made two lists to isolate the "Ask HN" and "Show HN" posts. We show that there are 1,744 "Ask HN" posts and 1,162 "Show HN" posts. In the next steps, we'll calculate the average comments for both "Ask HN" and "Show HN".

In [4]:
total_ask_comments = 0
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
    
avg_ask_comments = total_ask_comments / len(ask_posts)
print('Average comments per ASK post:', avg_ask_comments)

Average comments per ASK post: 14.038417431192661


In [5]:
total_show_comments = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments
    
avg_show_comments = total_show_comments / len(show_posts)
print('Average comments per SHOW post:', avg_show_comments)

Average comments per SHOW post: 10.31669535283993


On average, posts that begin with 'Ask HN' receive more comments. Going forward, we'll focus on "Ask HN" posts to find the best times to post that will generate the most comments. To do this, we'll use the 'created_at' column to extract the hour.

In [6]:
import datetime as dt

In [7]:
result_list = []
for row in ask_posts:
    created_date = row[6]
    num_comments = int(row[4])
    result_list.append([created_date, num_comments])
        
print(result_list[:5])

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


In [8]:
counts_by_hour = {} 
#number of ask posts created during each hour of the day.

comments_by_hour = {} 
#corresponding number of comments ask posts created at each hour received.

for row in result_list:
    hour = row[0]
    comment = row[1]
    time = dt.datetime.strptime(hour, "%m/%d/%Y %H:%M").strftime("%H")
    if time in counts_by_hour:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comment
    else:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comment

print('Total Comments by Hour:')        
comments_by_hour


Total Comments by Hour:


{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

In [9]:
avg_by_hour = []

for time in comments_by_hour:
    avg_by_hour.append([comments_by_hour[time] / counts_by_hour[time], time])
    
avg_by_hour = sorted(avg_by_hour, reverse=True)


# alternatively, we could take these extra steps for the same result.
# original code had the columns swapped
# 
# swap_avg_by_hour = []
# for row in avg_by_hour:
#     swap_avg_by_hour.append([row[1], row[0]])    
# sorted_swap = sorted(swap_avg_by_hour, reverse=True)
# sorted_swap

print('Top 5 Hours for Ask Posts Comments:')
print("(Eastern Time, US)")
print('\n')
for row in avg_by_hour[:5]:
    avg = row[0]
    time = row[1]
    time = dt.datetime.strptime(time, '%H').strftime('%H:%M')
    print(("{}: {:,.2f} average comments per post").format(time, avg)) 

Top 5 Hours for Ask Posts Comments:
(Eastern Time, US)


15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The the times that generate the most responses in the "Ask HN" category is 3PM, 2AM, 8PM, 4PM and 9PM. 